File: packages/butler-agent/src/gateways/app/domain/sessions/turn-cancellation.ts
import { Database } from "bun:sqlite";
import { AppStoreOperationError } from
  "../../infrastructure/core/app-store-errors.ts";
import type { TurnRow } from "../../infrastructure/core/records.ts";
import type {
  MessageRecord,
  TurnActionResult,
  TurnRecord,
} from "../../interface/protocol/app-protocol.ts";
import type { ButlerServiceClient } from "../../../core/client.ts";
import { sessionHintForRow } from "./session-read-model.ts";

export type TurnCancellationInput = {
  db: Database;
  getTurn: (turnId: string) => TurnRecord;
  getTurnRow: (turnId: string) => TurnRow | null;
  serviceClient: ButlerServiceClient;
  finalizeCancelledTurn: (chatId: string, turnId: string) => TurnRecord;
  cleanupTurnEventSequences: (chatId: string, turnId: string) => void;
  ensureCancelledTurnActivityMessage: (
    chatId: string,
    turnId: string,
  ) => MessageRecord | null;
};

export class AppTurnCancellation {
  constructor(private readonly input: TurnCancellationInput) {}

  async request(turnId: string): Promise<TurnActionResult> {
    const row = this.requireCancellableTurn(turnId);
    if (row.state === "cancelled") {
      this.input.ensureCancelledTurnActivityMessage(row.chat_id, turnId);
      return emptyResult(this.input.getTurn(turnId));
    }

    this.recordDecision(row);
    this.dispatchPending(turnId, row.chat_id);
    return emptyResult(this.input.getTurn(turnId));
  }

  reconcile(sessionId?: string): void {
    const pending = this.input.db
      .query<{ id: string; chat_id: string }, [string | null, string | null]>(`
        SELECT turns.id, turns.chat_id
        FROM turns
        JOIN app_turn_cancel_outbox
          ON app_turn_cancel_outbox.turn_id = turns.id
        WHERE turns.state = 'cancelling'
          AND app_turn_cancel_outbox.state = 'pending'
          AND app_turn_cancel_outbox.queue_id IS NULL
          AND (? IS NULL OR turns.chat_id = ?)
        ORDER BY turns.rowid ASC
      `)
      .all(sessionId ?? null, sessionId ?? null);
    for (const row of pending) this.dispatchPending(row.id, row.chat_id);

    const rows = this.input.db
      .query<{ id: string; chat_id: string }, [string | null, string | null]>(`
        SELECT turns.id, turns.chat_id
        FROM turns
        JOIN app_turn_cancel_outbox
          ON app_turn_cancel_outbox.turn_id = turns.id
        WHERE turns.state = 'cancelling'
          AND app_turn_cancel_outbox.state = 'completed'
          AND (? IS NULL OR turns.chat_id = ?)
        ORDER BY turns.rowid ASC
      `)
      .all(sessionId ?? null, sessionId ?? null);
    for (const row of rows) {
      this.input.finalizeCancelledTurn(row.chat_id, row.id);
      this.input.cleanupTurnEventSequences(row.chat_id, row.id);
    }
  }

  private requireCancellableTurn(turnId: string): TurnRow {
    const row = this.input.getTurnRow(turnId);
    if (!row) {
      throw new AppStoreOperationError(404, "turn_not_found", "Turn not found.");
    }
    const resuming = row.state === "cancelling";
    if (
      row.state !== "cancelled" &&
      ((!row.cancellable && !resuming) ||
        ["delivered", "failed"].includes(row.state))
    ) {
      throw new AppStoreOperationError(
        409,
        "turn_not_cancellable",
        "Turn is not cancellable.",
      );
    }
    return row;
  }

  private recordDecision(row: TurnRow): void {
    const now = new Date().toISOString();
    this.input.db.transaction(() => {
      const decision = this.input.db.query(`
        UPDATE turns
        SET state = 'cancelling', safe_status_label = 'Stopping',
          safe_error_code = NULL, retryable = 0, cancellable = 0, updated_at = ?
        WHERE id = ?
          AND state NOT IN ('cancelled', 'delivered', 'failed', 'runtime_fault')
      `).run(now, row.id);
      if (decision.changes !== 1) {
        throw new AppStoreOperationError(
          409,
          "turn_not_cancellable",
          "Turn is not cancellable.",
        );
      }
      this.input.db.query(`
        INSERT INTO app_turn_cancel_outbox (
          turn_id, queue_id, dispatch_claim_id, state, created_at
        ) VALUES (?, NULL, NULL, 'pending', ?)
        ON CONFLICT(turn_id) DO NOTHING
      `).run(row.id, now);
    })();
  }

  private dispatchPending(turnId: string, chatId: string): void {
    const pending = this.input.db.query<{ created_at: string }, [string]>(`
      SELECT created_at FROM app_turn_cancel_outbox
      WHERE turn_id = ? AND state = 'pending' AND queue_id IS NULL
    `).get(turnId);
    if (!pending) return;
    const queueClaimId = this.input.db.query<{
      claim_id: string | null;
    }, [string, string]>(`
      SELECT claim_id
      FROM session_queued_messages
      WHERE chat_id = ? AND turn_id = ? AND state = 'dispatching'
    `).get(chatId, turnId)?.claim_id ?? undefined;
    const queued = this.input.serviceClient.enqueueAppCancellation({
      chatId,
      sessionId: sessionHintForRow(chatId),
      turnId,
      requestId: `cancel:${turnId}`,
      requestedAt: pending.created_at,
      ...(queueClaimId ? { appQueueClaimId: queueClaimId } : {}),
    }, { source: "app-turn-cancellation" });
    this.input.db.query(`
      UPDATE app_turn_cancel_outbox SET queue_id = ?
      WHERE turn_id = ? AND state = 'pending' AND queue_id IS NULL
    `).run(queued.queueId, turnId);
  }
}

function emptyResult(turn: TurnRecord): TurnActionResult {
  return { turn, replies: [], next_cursor: 0 };
}

File: packages/butler-agent/rust/agent/src/gateway/application/turn_cancellation.rs
use rusqlite::{OptionalExtension, params};
use serde_json::{Value, json};

use super::*;

impl AppApplication {
    pub(super) async fn recover_turn_cancellations(&self) -> Result<(), GatewayApplicationError> {
        let pending = self
            .storage
            .execute(|db| {
                let mut statement = db
                    .prepare(
                        "SELECT turns.chat_id,turns.id,outbox.created_at,q.claim_id FROM app_turn_cancel_outbox outbox JOIN turns ON turns.id=outbox.turn_id LEFT JOIN session_queued_messages q ON q.chat_id=turns.chat_id AND q.turn_id=turns.id AND q.state='dispatching' WHERE outbox.state='pending' AND outbox.queue_id IS NULL ORDER BY turns.rowid",
                    )
                    .map_err(AppStorageError::sqlite)?;
                let rows = statement
                    .query_map([], |row| {
                        Ok((
                            row.get::<_, String>(0)?,
                            row.get::<_, String>(1)?,
                            row.get::<_, String>(2)?,
                            row.get::<_, Option<String>>(3)?,
                        ))
                    })
                    .map_err(AppStorageError::sqlite)?
                    .collect::<Result<Vec<_>, _>>()
                    .map_err(AppStorageError::sqlite)?;
                Ok(rows)
            })
            .await
            .map_err(app_error)?;
        for (chat, turn, requested_at, claim) in pending {
            self.dispatch_cancel(chat, turn, requested_at, claim)
                .await?;
        }
        Ok(())
    }

    pub(super) async fn cancel_turn_owned(
        &self,
        turn_id: String,
    ) -> Result<Value, GatewayApplicationError> {
        let requested_at = self.dependencies.identity_clock.now_iso();
        let stored_at = requested_at.clone();
        let turn = turn_id.clone();
        let prepared = self
            .storage
            .execute(move |db| {
                let row = db
                    .query_row(
                        "SELECT chat_id,state,cancellable FROM turns WHERE id=?1",
                        [&turn],
                        |row| {
                            Ok((
                                row.get::<_, String>(0)?,
                                row.get::<_, String>(1)?,
                                row.get::<_, bool>(2)?,
                            ))
                        },
                    )
                    .optional()
                    .map_err(AppStorageError::sqlite)?
                    .ok_or_else(|| {
                        AppStorageError::new("turn_not_found", "Turn not found.")
                    })?;
                if row.1 == "cancelled" {
                    return Ok((row.0, None, true));
                }
                if (!row.2 && row.1 != "cancelling")
                    || matches!(row.1.as_str(), "delivered" | "failed" | "runtime_fault")
                {
                    return Err(AppStorageError::new(
                        "turn_not_cancellable",
                        "Turn is not cancellable.",
                    ));
                }
                let tx = db.transaction().map_err(AppStorageError::sqlite)?;
                tx.execute(
                    "UPDATE turns SET state='cancelling',safe_status_label='Stopping',safe_error_code=NULL,retryable=0,cancellable=0,updated_at=?1 WHERE id=?2 AND state NOT IN ('cancelled','delivered','failed','runtime_fault')",
                    params![stored_at, turn],
                )
                .map_err(AppStorageError::sqlite)?;
                tx.execute(
                    "INSERT INTO app_turn_cancel_outbox(turn_id,queue_id,dispatch_claim_id,state,created_at) VALUES(?1,NULL,NULL,'pending',?2) ON CONFLICT(turn_id) DO NOTHING",
                    params![turn, stored_at],
                )
                .map_err(AppStorageError::sqlite)?;
                let queued = tx
                    .query_row(
                        "SELECT queue_id FROM app_turn_cancel_outbox WHERE turn_id=?1",
                        [&turn],
                        |row| row.get::<_, Option<String>>(0),
                    )
                    .map_err(AppStorageError::sqlite)?;
                let claim = tx
                    .query_row(
                        "SELECT claim_id FROM session_queued_messages WHERE chat_id=?1 AND turn_id=?2 AND state='dispatching'",
                        params![row.0, turn],
                        |row| row.get::<_, Option<String>>(0),
                    )
                    .optional()
                    .map_err(AppStorageError::sqlite)?
                    .flatten();
                tx.commit().map_err(AppStorageError::sqlite)?;
                Ok((row.0, claim, queued.is_some()))
            })
            .await
            .map_err(app_error)?;
        if !prepared.2 {
            self.dispatch_cancel(
                prepared.0.clone(),
                turn_id.clone(),
                requested_at,
                prepared.1,
            )
            .await?;
        }
        let turn = self
            .storage
            .execute(move |db| read_model::exact_turn(db, &turn_id))
            .await
            .map_err(app_error)?
            .ok_or_else(|| public(404, "turn_not_found", "Turn not found."))?;
        Ok(json!({"turn":turn,"replies":[],"next_cursor":0}))
    }

    async fn dispatch_cancel(
        &self,
        chat_id: String,
        turn_id: String,
        requested_at: String,
        app_queue_claim_id: Option<String>,
    ) -> Result<(), GatewayApplicationError> {
        let receipt = self
            .dependencies
            .native_ingress
            .enqueue_cancel(NativeAppCancellation {
                session_id: app_session_hint(&chat_id),
                chat_id,
                request_id: format!("cancel:{turn_id}"),
                turn_id: turn_id.clone(),
                requested_at,
                app_queue_claim_id,
            })
            .await?;
        self.storage
            .execute(move |db| {
                db.execute(
                    "UPDATE app_turn_cancel_outbox SET queue_id=?1 WHERE turn_id=?2 AND state='pending' AND queue_id IS NULL",
                    params![receipt.queue_id, turn_id],
                )
                .map_err(AppStorageError::sqlite)?;
                Ok(())
            })
            .await
            .map_err(app_error)
    }
}
